velox/functions/prestosql/json/JsonStringUtil.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <array>
+
+#include "folly/Unicode.h"
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/functions/prestosql/json/JsonStringUtil.h"
+
+namespace facebook::velox {
+namespace {
+/// Return the length in byte of the next UTF-8 encoded character at the
+/// beginning of `string`. If the beginning of `string` is not valid UTF-8
+/// encoding, return -1.
+FOLLY_ALWAYS_INLINE size_t validateAndGetNextUtf8Length(
+    const unsigned char* string,
+    const unsigned char* end) {
+  if (string == end) {
+    return 0;
+  }
+  if ((*string & 0x80u) == 0) {
+    return 1;
+  }
+  if ((*string & 0xE0u) == 0xC0u && (string + 1) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u) {
+    return 2;
+  }
+  if ((*string & 0xF0u) == 0xE0u && (string + 2) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u && (*(string + 2) & 0xC0u) == 0x80u) {
+    return 3;
+  }
+  if ((*string & 0xF8u) == 0xF0u && (string + 3) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u && (*(string + 2) & 0xC0u) == 0x80u &&
+      (*(string + 3) & 0xC0u) == 0x80u) {
+    return 4;
+  }
+  return -1;
+}
+
+FOLLY_ALWAYS_INLINE char hexDigit(uint8_t c) {
+  return c < 10 ? c + '0' : c - 10 + 'A';
+}
+
+FOLLY_ALWAYS_INLINE void writeHex(char16_t value, char*& out) {
+  *out++ = '\\';
+  *out++ = 'u';
+  *out++ = hexDigit((value >> 12) & 0x0F);
```

Comment:
Add `value = folly::Endian::little(value)` to be safe

Reply:
Added. Good to know.